### ISSUE TITLE
build fails: error in format-handler

### ISSUE THREAD

**Issue by BalticPinguin** [libMesh/libmesh#3589]
I recently updated my toolchain and ran into a compilation error.
I configure libMesh with  `  --enable-werror   --enable-petsc `  (I consider those to be the relevant flags) and during `make` I got the error
````
../master/src/solvers/petscdmlibmeshimpl.C: In function ‘PetscErrorCode DMCreateDomainDecomposition_libMesh(DM, PetscInt*, char***, _p_IS***, _p_IS***, _p_DM***)’:
../master/src/solvers/petscdmlibmeshimpl.C:447:85: error: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘long unsigned int’ [-Werror=format=]
  447 |         if (elen != dlen) LIBMESH_SETERRQ1(((PetscObject)dm)->comm, PETSC_ERR_PLIB, "Failed to embed field %" LIBMESH_PETSCINT_FMT, d);
      |                                                                                                                                     ~
      |                                                                                                                                     |
      |                                                                                                                                     long unsigned int
/data/tm162/bin/DSK/petsc_opt/include/petscerror.h:57:127: note: in definition of macro ‘SETERRQ’
   57 |     PetscErrorCode ierr_seterrq_petsc_ = PetscError(comm, __LINE__, PETSC_FUNCTION_NAME, __FILE__, ierr, PETSC_ERROR_INITIAL, __VA_ARGS__); \
      |                                                                                                                               ^~~~~~~~~~~
../master/src/solvers/petscdmlibmeshimpl.C:447:27: note: in expansion of macro ‘LIBMESH_SETERRQ1’
  447 |         if (elen != dlen) LIBMESH_SETERRQ1(((PetscObject)dm)->comm, PETSC_ERR_PLIB, "Failed to embed field %" LIBMESH_PETSCINT_FMT, d);
      |                           ^~~~~~~~~~~~~~~~
In file included from /data/tm162/bin/DSK/petsc_opt/include/petscsys.h:44,
                 from /data/tm162/bin/DSK/petsc_opt/include/petscbag.h:5,
                 from /data/tm162/bin/DSK/petsc_opt/include/petsc.h:5,
                 from ./include/libmesh/petsc_macro.h:66,
                 from ../master/src/solvers/petscdmlibmeshimpl.C:20:
/data/tm162/bin/DSK/petsc_opt/include/petscsystypes.h:293:25: note: format string is defined here
  293 |   #define PetscInt_FMT "d"
cc1plus: all warnings being treated as errors
````
According to my analysis, the problem is that the variable `d` is supposed to be printed using `%d` ; but it is defined ( in ../master/src/solvers/petscdmlibmeshimpl.C:395) as
`for (auto d : index_range(*dlm->decomposition)) {`, which is of type  `size_type`, which is [Unsigned integer type (usually std::size_t)](https://en.cppreference.com/w/cpp/container/vector).

I am not sure how to handle this issue in best case: First, for some reason I seem to be the only one having trouble with this, which makes me wonder.
And there are [lengthy discussions](https://stackoverflow.com/questions/1546789/clean-code-to-printf-size-t-in-c-or-nearest-equivalent-of-c99s-z-in-c) about ways to handle such a situation. Since I don't see a simple and elegant variant, I'd like to hear some opinions...

**Comment by roystgnr** [comment]
What version of PETSc are you using?

**Comment by BalticPinguin** [comment]
I use the current git-version: It counts as 3.19.2.

**Comment by jwpeterson** [comment]
Looks like the same type of issue that was fixed in libMesh/libMesh@e3ca53c3840a9c53eb05ce2dffc976b99da5f59d

**Comment by BalticPinguin** [comment]
yes; essentially, the first two fixes (lines 345 and 446) are what cause the trouble for me (Above, only the latter is posted; but in fact the same error is listed for the statement in line 345.
In my case, the compiler (gcc 11.2.0) is not happy because `LIBMESH_PETSCINT_FMT` is `d`:
````
/data/tm162/bin/DSK/petsc_opt/include/petscsystypes.h:293:25: note: format string is defined here
  293 |   #define PetscInt_FMT "d"
````
The reference to #3417 is quite useful: I just don't see why it worked for them, but not for me.
It should be an `unsigned int` in both cases!?

**Comment by jwpeterson** [comment]
```
#if PETSC_RELEASE_LESS_THAN(3,8,0)
# define LIBMESH_PETSCINT_FMT "D"
#else
# define LIBMESH_PETSCINT_FMT PetscInt_FMT
#endif
```

I think the issue is that you are not using a "release version". Since you are using the current git version, PETSC_RELEASE_LESS_THAN(x,y,z) is always false for your build, regardless of what (x,y,z) are, because your build is not a release.

**Comment by jwpeterson** [comment]
Hmm... actually I guess git version vs. release version doesn't make a difference in this case, because even if the code was using `PETSC_VERSION_LESS_THAN(3,8,0)` it would still be false for you and you would end up with `PetscInt_FMT`. So, disregard the previous comment, I think they must have really changed something in 3.19.2 that breaks our code.

**Comment by roystgnr** [comment]
I'll see if I can replicate with 3.19.2.  If something's been broken in between release tags, though, you're on your own; it's hard enough to keep up with compatibility issues coming up in between subminor releases...

**Comment by BalticPinguin** [comment]
Thank you very much for your help!
I'll also see if it works with version 3.18.

For the record: The `PETSC_RELEASE_LESS_THAN(x,y,z)` works also for git-versions:
In my  `libmesh_config.h` I can find
````
/* PETSc's major version number, as detected by petsc.m4 */
#ifndef LIBMESH_DETECTED_PETSC_VERSION_MAJOR
#define LIBMESH_DETECTED_PETSC_VERSION_MAJOR 3
#endif

/* PETSc's minor version number, as detected by petsc.m4 */
#ifndef LIBMESH_DETECTED_PETSC_VERSION_MINOR
#define LIBMESH_DETECTED_PETSC_VERSION_MINOR 19
#endif

/* PETSc release (1) or petsc-dev (0), as detected by petsc.m4 */
#ifndef LIBMESH_DETECTED_PETSC_VERSION_RELEASE
#define LIBMESH_DETECTED_PETSC_VERSION_RELEASE 0
#endif

/* PETSc's subminor version number, as detected by petsc.m4 */
#ifndef LIBMESH_DETECTED_PETSC_VERSION_SUBMINOR
#define LIBMESH_DETECTED_PETSC_VERSION_SUBMINOR 2
#endif
````

**Comment by jwpeterson** [comment]
Yeah, the key point is that:
```
#define LIBMESH_DETECTED_PETSC_VERSION_RELEASE 0
```
is false, so the second part of the macro:
```
#define PETSC_RELEASE_LESS_THAN(major,minor,subminor)                   \
  (PETSC_VERSION_LESS_THAN(major,minor,subminor) && LIBMESH_DETECTED_PETSC_VERSION_RELEASE)
```
is false for your build.

**Comment by roystgnr** [comment]
That ... was not what I was expecting when I picked that macro over `PETSC_VERSION_LESS_THAN`.  But it would explain why I'm not able to replicate the problem on 3.19.2.

What is the point of `PETSC_RELEASE_LESS_THAN`, exactly?  `PETSC_VERSION_LESS_THAN` defines a connected interval of "PETSc commits before the version number was changed", which isn't going to be the same as "PETSc commits before the API was changed" but which ought to be pretty close.  But if `PETSC_RELEASE_LESS_THAN` defines a discrete set of "PETSc releases before the version number was changed, but not the version control commits between those releases", isn't that basically asking to not be compatible with any non-release PETSc?

**Comment by jwpeterson** [comment]
I think the main point of `PETSC_RELEASE_LESS_THAN` is for it to be temporary, e.g.

```
#if PETSC_RELEASE_LESS_THAN(3,18,0)
// code that works for released PETSc 3.17.x and older
#else 
// code that is in a new-but-not-yet-released git-version of PETSc 3.18.x
#endif 
```
Of course it's temporary since 3.18.1 or whatever will eventually be released, once again breaking the code.

**Comment by roystgnr** [comment]
Okay, that makes sense.  But then it's not just me - all 10 of our current use cases are basically atavisms that we should switch to `PETSC_VERSION_LESS_THAN`?

**Comment by jwpeterson** [comment]
> all 10 of our current use cases are basically atavisms that we should switch to `PETSC_VERSION_LESS_THAN`?

I think that's right, yeah.

**Comment by roystgnr** [comment]
I'm still not seeing any warnings with g++ 11.3 on 98340805 (the Jun 21 PETSc hash)...

**Comment by roystgnr** [comment]
Okay, we've got a hypothesis: I'm building with 64-bit indices; you're probably not, and with 64-bit indices `PetscInt_FMT` and `PetscInt64_FMT` are the same, but with 32-bit indices we're passing a `size_t` to a 32-bit format character from `PetscInt_FMT` and that's not kosher.

**Comment by roystgnr** [comment]
Yup, that was it.  32-bit build and I can replicate the warning.  We only test 32-bit in opt mode (which has almost no warnings) and without -Werror in CI; I'll fix that shortly after I fix this.

**Comment by roystgnr** [comment]
Check if #3592 works for you, and let me know?

**Comment by roystgnr** [comment]
I merged since that PR fixed things for me at least, but if you still hit the problem let me know and we'll reopen.

**Comment by BalticPinguin** [comment]
Great, thank you very much for the fast fix, while I was still evaluating that also the 3.18-variant didn't work ;-)

But the fix #3592 indeed works for me. Thank you very much!
